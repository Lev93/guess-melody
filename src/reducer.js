/* eslint-disable arrow-body-style */
const initialState = {
  step: -1,
  mistakes: 0,
};

const ActionType = {
  INCREMENT_MISTAKES: 'INCREMENT_MISTAKES',
  INCREMENT_STEP: 'INCREMENT_STEP',
  RESET: 'RESET',
};

const isArtistAnswerCorrect = (userAnswer, question) => userAnswer.artist === question.song.artist;

const isGenreAnswerCorrect = (userAnswer, question) => userAnswer.every((it, i) => it === (
  question.answers[i].genre === question.genre
));

const ActionCreator = {
  incrementStep: () => ({
    type: ActionType.INCREMENT_STEP,
    payload: 1,
  }),

  incrementMistake: (userAnswer, question) => {
    let answerIsCorrect = false;

    switch (question.type) {
      case 'artist':
        answerIsCorrect = isArtistAnswerCorrect(userAnswer, question);
        break;
      case 'genre':
        answerIsCorrect = isGenreAnswerCorrect(userAnswer, question);
        break;
      default: console.log('unknown question');
    }

    return {
      type: ActionType.INCREMENT_MISTAKES,
      payload: answerIsCorrect ? 0 : 1,
    };
  },
  resetGame: () => {
    return {
      type: ActionType.RESET,
    };
  },
};


const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'INCREMENT_STEP': {
      return { ...state, step: state.step + action.payload };
    }
    case 'INCREMENT_MISTAKES': {
      return { ...state, mistakes: state.mistakes + action.payload };
    }

    case 'RESET': return { ...initialState };
    default:
      return state;
  }
};


export {
  ActionCreator,
  ActionType,
  isArtistAnswerCorrect,
  isGenreAnswerCorrect,
  reducer,
};
