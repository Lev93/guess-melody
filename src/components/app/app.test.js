import React from 'react';
import renderer from 'react-test-renderer';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { reducer } from '../../reducer';

import App from './app.jsx';

const mock = {
  questions: [
    {
      type: 'genre',
      genre: 'rock',
      answers: [
        {
          src: 'test.mp3',
          genre: 'rock',
        },
        {
          src: 'test.mp3',
          genre: 'blues',
        },
        {
          src: 'test.mp3',
          genre: 'jazz',
        },
        {
          src: 'test.mp3',
          genre: 'rock',
        },
      ],
    },
    {
      type: 'artist',
      song: {
        artist: 'Jim Beam',
        src: 'path.mp3',
      },
      answers: [
        {
          picture: 'path.jpg',
          artist: 'John Snow',
        },
        {
          picture: 'path.jpg',
          artist: 'Jack Daniels',
        },
        {
          picture: 'path.jpg',
          artist: 'Jim Beam',
        },
      ],
    },
  ],
};


it('App correctly renders first screen', () => {
  const { questions } = mock;
  const store = createStore(reducer);
  const tree = renderer.create(<Provider store={store}>
    <App
    mistakes={0}
    maxMistakes={Infinity}
    gameTime={100}
    questions={questions}
    step={-1}
    onUserAnswer={jest.fn()}
    onWelcomeScreenClick={jest.fn()}
  /></Provider>).toJSON();

  expect(tree).toMatchSnapshot();
});

it('App correctly renders genre question screen', () => {
  const { questions } = mock;
  const store = createStore(reducer);
  const tree = renderer.create(<Provider store={store}><App
    mistakes={0}
    maxMistakes={Infinity}
    gameTime={100}
    questions={questions}
    step={1}
    onUserAnswer={jest.fn()}
    onWelcomeScreenClick={jest.fn()}
  /></Provider>, {
    createNodeMock: () => {},
  }).toJSON();

  expect(tree).toMatchSnapshot();
});

it('App correctly renders artist question screen', () => {
  const { questions } = mock;
  const store = createStore(reducer);
  const tree = renderer.create(<Provider store={store}><App
    mistakes={0}
    maxMistakes={Infinity}
    gameTime={100}
    questions={questions}
    step={2}
    onUserAnswer={jest.fn()}
    onWelcomeScreenClick={jest.fn()}
  /></Provider>, {
    createNodeMock: () => {},
  }).toJSON();

  expect(tree).toMatchSnapshot();
});
