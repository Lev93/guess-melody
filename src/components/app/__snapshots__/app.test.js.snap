// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App correctly renders artist question screen 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <div
      className="timer__value"
    >
      <span
        className="timer__mins"
      >
        05
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        00
      </span>
    </div>
    <div
      className="game__mistakes"
    >
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
    </div>
  </header>
</section>
`;

exports[`App correctly renders first screen 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <div
      className="timer__value"
    >
      <span
        className="timer__mins"
      >
        05
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        00
      </span>
    </div>
    <div
      className="game__mistakes"
    >
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
    </div>
  </header>
</section>
`;

exports[`App correctly renders genre question screen 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <div
      className="timer__value"
    >
      <span
        className="timer__mins"
      >
        05
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        00
      </span>
    </div>
    <div
      className="game__mistakes"
    >
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
    </div>
  </header>
</section>
`;
